lib/parse/modelFactory.js

```diff
@@ -2,149 +2,140 @@ var Graph = require('../model/graph');
 var Edge = require('../model/edge');
 var Vertex = require('../model/vertex');
 var Port = require('../model/port');
-var CardinalPortPosition = require('../model/cardinalPortPosition');
 var PortType = require('../enum/portType');
 
 var ObjectUtils = require('../utilities/objects');
 var Lang = require('../utilities/lang');
 
-var setIfNotExist = function(id, obj, map) {
-  if (Object.keys(map).indexOf(id) !== -1) {
-    console.error('Duplicate entity id: ' + id);
-    return false;
-  }
-  map[id] = obj;
-  return true;
-};
-
-
-var SCOPE_SYMBOLS = {
-  VERTEX: '>',
-  INPUT: '+',
-  OUTPUT: '-'
-};
-
-var INVALID_ID_CHARS = ObjectUtils.values(SCOPE_SYMBOLS);
-
 /**
- * Creates data model objects from an abstract syntax tree with some validation.
- *
  * @class ModelFactory
+ * @static
  */
-var ModelFactory = function() {
-  // all built models keyed by global id
-  this._vertices = {};
-  this._edges = {};
-  this._ports = {};
-};
-
-ModelFactory.prototype = {
-  /**
-   * Parse an abstract syntax tree into a Graph model
-   * @method buildGraphModel
-   * @param  {Object} data - the AST
-   * @returns {Graph}
-   */
-  buildGraphModel: function(data) {
-    data.vertices.forEach(function(vertexData) {
-      this.parseVertex(vertexData);
-    }.bind(this));
+var ModelFactory = {
+
+  buildGraph: function(data, parent) {
+    var parentId = null;
+    if (!(parent instanceof Vertex)) {
+      parent = null;
+    } else {
+      parentId = parent.get('globalId');
+    }
 
-    data.edges.forEach(function(edgeData) {
-      this.parseEdge(edgeData);
-    }.bind(this));
+    var vertices = data.vertices;
+    var edges = data.edges;
 
-    /* note: we maintain state to validate that ids are unique.
-      a simple stateless mapping function would allow duplicate ids
-      */
+    var vertexModels = {};
+    ObjectUtils.each(vertices, function(vertexData, id) {
+      vertexModels[id] = ModelFactory.buildVertex(id, vertexData, parentId);
+    }, this);
 
-    var vertices = ObjectUtils.values(this._vertices);
-    var edges = ObjectUtils.values(this._edges);
+    var edgeModels = {};
+    ObjectUtils.each(edges, function(edgeData, id) {
+      edgeModels[id] = ModelFactory.buildEdge(id, edgeData, vertexModels, parent);
+    });
 
     return new Graph({
-      vertices: vertices,
-      edges: edges
+      vertices: ObjectUtils.values(vertexModels),
```

Comment:
okay so this is a bit silly because the constructor for Graph rebulids a map of ids to vertices after we flatten such a map to an array here.